pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go

```diff
@@ -0,0 +1,241 @@
+				log.Error(err, err.Error(), "nsxserviceaccount", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultNormal, nil
+			}
+			log.Error(err, "operate failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResType)
+			if err := r.Service.DeleteNSXServiceAccount(ctx, types.NamespacedName{
+				Namespace: obj.Namespace,
+				Name:      obj.Name,
+			}); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			controllerutil.RemoveFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "nsxserviceaccount", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			// only print a message because it's not a normal case
+			log.Info("finalizers cannot be recognized", "nsxserviceaccount", req.NamespacedName)
+		}
+	}
+
+	return ResultNormal, nil
+}
+
+// setupWithManager sets up the controller with the Manager.
+func (r *NSXServiceAccountReconciler) setupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&nsxvmwarecomv1alpha1.NSXServiceAccount{}).
+		WithEventFilter(predicate.Funcs{
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				// Suppress Delete events to avoid filtering them out in the Reconcile function
+				return false
+			},
+		}).
+		WithOptions(
+			controller.Options{
+				MaxConcurrentReconciles: runtime.NumCPU(),
+			}).
+		Complete(r)
+}
+
+// Start setup manager and launch GC
+func (r *NSXServiceAccountReconciler) Start(mgr ctrl.Manager) error {
+	err := r.setupWithManager(mgr)
+	if err != nil {
+		return err
+	}
+
+	go r.GarbageCollector(make(chan bool), servicecommon.GCInterval)
+	return nil
+}
+
+// GarbageCollector collect NSXServiceAccount which has been removed from crd.
+// cancel is used to break the loop during UT
+func (r *NSXServiceAccountReconciler) GarbageCollector(cancel chan bool, timeout time.Duration) {
+	ctx := context.Background()
+	log.Info("garbage collector started")
+	for {
+		select {
+		case <-cancel:
+			return
+		case <-time.After(timeout):
+		}
+		nsxServiceAccountUIDSet := r.Service.ListNSXServiceAccountRealization()
+		if len(nsxServiceAccountUIDSet) == 0 {
+			continue
+		}
+		nsxServiceAccountList := &nsxvmwarecomv1alpha1.NSXServiceAccountList{}
+		err := r.Client.List(ctx, nsxServiceAccountList)
```

Comment:
Is it possible to use a cache for NSXServiceAccount, and update the cache in sync with CR Added/Updated/Deleted events? Although NSXServiceAccount object is small, I feel better to avoid a regular list call for all NSXServiceAccounts.

Reply:
The cache is used in kubebuilder.
https://book.kubebuilder.io/architecture.html

Reply:
Got it. Thanks. Found explanations in controller-runtime package doc.